syntax = "proto3";

package subscription.v1;

import "google/protobuf/timestamp.proto";

// Subscription Service - gRPC API
// Used for quota checking, subscription management, and entitlement verification
// Port: 50052
service SubscriptionService {
  // Check if user can create more resources (strategies, templates, etc.)
  rpc CheckResourceQuota(CheckResourceQuotaRequest) returns (CheckResourceQuotaResponse);

  // Check feature access (e.g., private_templates)
  rpc CheckFeatureAccess(CheckFeatureAccessRequest) returns (CheckFeatureAccessResponse);

  // Get user's subscription details
  rpc GetSubscription(GetSubscriptionRequest) returns (GetSubscriptionResponse);

  // Get user's entitlements (features and limits)
  rpc GetEntitlements(GetEntitlementsRequest) returns (GetEntitlementsResponse);

  // Get current usage for a specific metric (legacy: backtests, API calls, etc.)
  rpc GetUsage(GetUsageRequest) returns (GetUsageResponse);

  // Get all quota statuses for a user (legacy)
  rpc GetAllQuotas(GetAllQuotasRequest) returns (GetAllQuotasResponse);

  // Check if user can perform action within quota (legacy)
  rpc CheckQuota(CheckQuotaRequest) returns (CheckQuotaResponse);

  // Health check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// Request Messages

// CheckResourceQuotaRequest defines the request for resource quota check
message CheckResourceQuotaRequest {
  // User ID for authorization (auto-injected from metadata)
  string user_id = 1;
  // Resource type (strategies, strategy_versions_per_strategy, etc.)
  string resource_type = 2;
  // Current count of resources owned by user
  int32 current_count = 3;
}

// CheckFeatureAccessRequest defines the request for feature access check
message CheckFeatureAccessRequest {
  // User ID for authorization (auto-injected from metadata)
  string user_id = 1;
  // Feature name (private_templates, custom_indicators, etc.)
  string feature_name = 2;
}

// CheckQuotaRequest defines the request payload for CheckQuota (legacy)
message CheckQuotaRequest {
  // User ID for authorization
  string user_id = 1;
  // Metric to check (api_calls, backtests, ai_chat_messages, ai_tokens, storage_bytes)
  string metric = 2;
  // Amount to check (default: 1)
  int32 amount = 3;
}

// GetSubscriptionRequest defines the request payload for GetSubscription
message GetSubscriptionRequest {
  // User ID for authorization
  string user_id = 1;
}

// GetEntitlementsRequest defines the request payload for GetEntitlements
message GetEntitlementsRequest {
  // User ID for authorization
  string user_id = 1;
}

// GetUsageRequest defines the request payload for GetUsage
message GetUsageRequest {
  // User ID for authorization
  string user_id = 1;
  // Metric to get usage for
  string metric = 2;
  // Optional: specific date (YYYY-MM-DD, default: today)
  optional string date = 3;
}

// GetAllQuotasRequest defines the request payload for GetAllQuotas
message GetAllQuotasRequest {
  // User ID for authorization
  string user_id = 1;
  // Optional: specific date (YYYY-MM-DD, default: today)
  optional string date = 2;
}

// HealthCheckRequest defines the request payload for HealthCheck
message HealthCheckRequest {}

// Response Messages

// CheckResourceQuotaResponse defines the response for resource quota check
message CheckResourceQuotaResponse {
  // Whether user can create more resources
  bool can_create = 1;
  // Current resource count
  int32 current_count = 2;
  // Resource limit (0 = unlimited)
  int32 limit = 3;
  // Remaining quota (0 = unlimited)
  int32 remaining = 4;
  // Tier name
  string tier = 5;
  // Optional message (e.g., upgrade suggestion)
  optional string message = 6;
}

// CheckFeatureAccessResponse defines the response for feature access check
message CheckFeatureAccessResponse {
  // Whether feature is enabled
  bool enabled = 1;
  // Feature value (could be int limit, bool flag, or string config)
  oneof value {
    // Integer value (e.g., private_templates limit count)
    int32 int_value = 2;
    // Boolean value (e.g., optimization enabled/disabled)
    bool bool_value = 3;
    // String value (e.g., feature configuration)
    string string_value = 4;
  }
  // Tier name
  string tier = 5;
  // Optional message (e.g., upgrade suggestion)
  optional string message = 6;
}

// CheckQuotaResponse defines the response payload for CheckQuota (legacy)
message CheckQuotaResponse {
  // Whether the action is allowed
  bool allowed = 1;
  // Current usage for the metric
  int64 current_usage = 2;
  // Limit for the metric (0 = unlimited)
  int64 limit = 3;
  // Remaining quota (0 = unlimited)
  int64 remaining = 4;
  // Usage percentage (0-100+)
  double percentage = 5;
  // Quota status (allowed, warning, exceeded)
  string status = 6;
  // When quota resets
  google.protobuf.Timestamp reset_at = 7;
  // Optional message (e.g., reason for denial)
  optional string message = 8;
}

// GetSubscriptionResponse defines the response payload for GetSubscription
message GetSubscriptionResponse {
  // Subscription ID
  string id = 1;
  // User ID
  string user_id = 2;
  // Tier (free, starter, professional, institutional)
  string tier = 3;
  // Status (active, cancelled, expired, suspended)
  string status = 4;
  // Billing cycle (monthly, yearly)
  optional string billing_cycle = 5;
  // Price paid
  double price_paid = 6;
  // Currency (USD, EUR, etc.)
  string currency = 7;
  // Subscription start date
  google.protobuf.Timestamp started_at = 8;
  // Subscription expiry date
  optional google.protobuf.Timestamp expires_at = 9;
  // Cancellation date
  optional google.protobuf.Timestamp cancelled_at = 10;
  // Stripe subscription ID
  optional string stripe_subscription_id = 11;
  // Stripe customer ID
  optional string stripe_customer_id = 12;
  // Payment method
  optional string payment_method = 13;
  // Created at
  google.protobuf.Timestamp created_at = 14;
  // Updated at
  google.protobuf.Timestamp updated_at = 15;
}

// GetEntitlementsResponse defines the response payload for GetEntitlements
message GetEntitlementsResponse {
  // User ID
  string user_id = 1;
  // Tier name
  string tier = 2;
  // Feature access flags
  map<string, bool> features = 3;
  // Quota limits (0 = unlimited)
  QuotaLimits limits = 4;
}

// QuotaLimits defines quota limits for all metrics
message QuotaLimits {
  // API calls per day (0 = unlimited)
  int64 api_calls = 1;
  // Backtests per day (0 = unlimited)
  int64 backtests = 2;
  // AI chat messages per day (0 = unlimited)
  int64 ai_chat_messages = 3;
  // AI tokens per day (0 = unlimited)
  int64 ai_tokens = 4;
  // Storage bytes (0 = unlimited)
  int64 storage_bytes = 5;
}

// GetUsageResponse defines the response payload for GetUsage
message GetUsageResponse {
  // User ID
  string user_id = 1;
  // Usage date
  string date = 2;
  // Metric name
  string metric = 3;
  // Usage amount
  int64 usage = 4;
}

// GetAllQuotasResponse defines the response payload for GetAllQuotas
message GetAllQuotasResponse {
  // User ID
  string user_id = 1;
  // Quota statuses for all metrics
  repeated QuotaStatus quotas = 2;
}

// QuotaStatus defines quota status for a specific metric
message QuotaStatus {
  // Metric name
  string metric = 1;
  // Current usage
  int64 current_usage = 2;
  // Limit (0 = unlimited)
  int64 limit = 3;
  // Remaining quota (0 = unlimited)
  int64 remaining = 4;
  // Usage percentage (0-100+)
  double percentage = 5;
  // Quota status (allowed, warning, exceeded)
  string status = 6;
  // When quota resets
  google.protobuf.Timestamp reset_at = 7;
}

// HealthCheckResponse defines the response payload for HealthCheck
message HealthCheckResponse {
  // Service status
  string status = 1;
  // Service version
  string version = 2;
}
